server/src/main/java/org/elasticsearch/common/util/SerializableBytesRefHash.java

```diff
@@ -0,0 +1,214 @@
+        IntArray hashes,
+        int seed,
+        BigArrays bigArrays
+    ) {
+        super(capacity, maxLoadFactor, size, maxSize, ids, startOffsets, bytes, hashes, bigArrays);
+        this.seed = seed;
+    }
+
+    public static SerializableBytesRefHash loadFromStream(StreamInput in, BigArrays bigArrays) throws IOException {
+        float maxLoadFactor = in.readFloat();
+        long size = in.readVLong();
+        long maxSize = in.readVLong();
+        long capacity = in.readLong();
+
+        LongArray ids = null;
+        LongArray startOffsets = null;
+        ByteArray bytes = null;
+        IntArray hashes = null;
+        SerializableBytesRefHash hash = null;
+
+        // we allocate big arrays so we have to `close` if we fail here or we'll leak them.
+        boolean success = false;
+
+        try {
+            ids = bigArrays.newLongArray(capacity, true);
+            for (int i = 0; i < capacity; ++i) {
+                ids.set(i, in.readVLong());
+            }
+
+            // there is just 1 way of serializing at the moment
+            SerializationType serializationType = SerializationType.fromType(in.readByte());
+            if (SerializationType.RAW_BUFFERS.equals(serializationType) == false) {
+                throw new IllegalArgumentException("unsupported type");
+            }
+
+            // startOffsets
+            long sizeOfStartOffsets = in.readVLong();
+            startOffsets = bigArrays.newLongArray(sizeOfStartOffsets, true);
+            for (int i = 0; i < sizeOfStartOffsets; ++i) {
+                startOffsets.set(i, in.readVLong());
+            }
+
+            // bytes
+            long sizeOfBytes = in.readVLong();
+            bytes = bigArrays.newByteArray(sizeOfBytes, true);
+
+            for (int i = 0; i < sizeOfBytes; ++i) {
+                bytes.set(i, in.readByte());
+            }
+
+            // hashes
+            long sizeOfHashes = in.readVLong();
+            hashes = bigArrays.newIntArray(sizeOfHashes, true);
+            for (int i = 0; i < sizeOfHashes; ++i) {
+                hashes.set(i, in.readZInt());
+            }
+
+            int seed = in.readInt();
+            hash = new SerializableBytesRefHash(capacity, maxLoadFactor, size, maxSize, ids, startOffsets, bytes, hashes, seed, bigArrays);
+            success = true;
+        } finally {
+            if (success == false) {
+                Releasables.close(ids, startOffsets, bytes, hashes);
+            }
+        }
+
+        return hash;
+    }
+
+    @Override
+    public long find(BytesRef key) {
+        return find(key, hash(key));
+    }
+
+    @Override
+    public long add(BytesRef key) {
+        return add(key, hash(key));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeFloat(maxLoadFactor);
+        out.writeVLong(size);
+        out.writeVLong(maxSize);
+
+        long capacity = capacity();
+        out.writeLong(capacity);
```

Comment:
Could this be `writeVLong` or is there some situation when `capacity` can be negative?